src/zkevm_specs/evm/execution/mul.py

```diff
@@ -0,0 +1,66 @@
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+from ...util import FQ
+
+
+def mul(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    is_mul = (Opcode.DIV - opcode) * (Opcode.MOD - opcode) * FQ(8).inv()
+    is_div = (opcode - Opcode.MUL) * (Opcode.MOD - opcode) * FQ(4).inv()
+    is_mod = (opcode - Opcode.MUL) * (opcode - Opcode.DIV) * FQ(8).inv()
+
+    pop1 = instruction.stack_pop()
+    pop2 = instruction.stack_pop()
+    push = instruction.stack_push()
+
+    # this part corresponds to witness assignment in the zkevm circuit
+    if is_mul == 1:
+        a = pop1
+        b = pop2
+        c = instruction.rlc_encode(0, 32)
+        d = push
+    elif is_div == 1:
+        d = pop1  # dividend
+        b = pop2  # divisor
+        a = push  # quotient
+        c = instruction.rlc_encode(d.int_value - b.int_value * a.int_value, 32)  # remainder
+    else:  # is_mod == 1
+        d = pop1  # dividend
+        b = pop2  # divisor
+        if b.int_value == 0:
+            c = d
+            a = instruction.rlc_encode(0, 32)
+        else:
+            c = push
+            a = instruction.rlc_encode((d.int_value - c.int_value) // b.int_value, 32)
```

Comment:
This witness assignement model looks to me like it will fall into a `PermutationError` easily.

Let's say we compile this circuit and the use-case used to compile it is `mul`.
We will have the following copy-constraints from the stack cell to the circuit cell:
- copy(from: stack_pop_1, to: a)
- copy(from: stack_pop_2, to: b)
- copy(from: d, to: stack_push)

But if we re-use the same circuit instance and we get opcode `DIV`, now we have:
- copy(from: stack_pop_1, to: d)
- copy(from: stack_pop_2, to: b)
- copy(from: a, to: stack_push)


As you can see, the copies are different, and so, the permutation argument will differ too. Ending up with a `PermutationError`, as the cells that were supposed to contain equal values do not depending on the opcode we recieve.

If I'm right on this (I could be wrong, missing context etc..) we need a more tricky system which basically uses flags (like `is_mul`, `is_div`..) multiplied by the stack pops and pushes and multiplies the values doing additions of the three terms for each cell copy. On that way, we can centralize the copies from the same cell origin/destination and flags conditionally change the witnessed values instead of the positions of the witness.

Reply:
So after discussing with @han0110 and @ed255 I think the root of my confusion was that the specs are written on a way on which it seems that you're doing lookups based on conditions. When, in reality, is the expression used for the lookup the one which expresses a condition.

[This](https://github.com/appliedzkp/zkevm-circuits/pull/362/files#diff-8e75d3ba3ed5df97506a778cccc15aac2454cf57bf44cc0c758d621a3d2d73d8L36-R65) diff makes it clear for me.

I think we should try to write the specs a bit more similar to the actual implementation (saving the differences). So that on that case, we could try to make clear that we don't conditionally lookup, but we make expressions that are conditional. And thus are being used in the lookups.

Reply:
Speaking to @ed255 I saw that is also difficult to re-write this on a way on which it does not seem that the lookups are kinda conditional.

The python API we have here returns a value for the lookup. And so will be difficult to reflect how the real circuit impl will look like. 

Let's leave it like this and maybe just add some comments in regards that.
Apologies for the noise :sweat_smile:

Reply:
@CPerezz I understand your confusion. It also took me a while to figure out a way to write the spec given the current Python API. The root cause is that we mix the witness assignment and constraints together, and the lookup returns the value instead of feeding the value into lookup.

So in addition to the assignment, I added extra [checks](https://github.com/appliedzkp/zkevm-circuits/pull/362/files#diff-8e75d3ba3ed5df97506a778cccc15aac2454cf57bf44cc0c758d621a3d2d73d8R61-R71) below to make sure that push/pop values match the `a, b, c, d` value used in the `mul_add_words` gadget, which also reflects the logic implemented in the rust [circuit](https://github.com/appliedzkp/zkevm-circuits/pull/362/files#diff-8e75d3ba3ed5df97506a778cccc15aac2454cf57bf44cc0c758d621a3d2d73d8R61-R71)

Reply:
Yep this looks perfect. But took me a while to understand what  you were doing indeed :) The rust code made it much more clear.

I think with the Python API we have, that's a really good definition already! :)